### PR TITLE
sock: change IPv4 address type to array

### PR DESCRIPTION
**Pull request by miri64**
I really had problems in #5936, #5937, and #5938 to keep byte order. To keep in line with the portability credo of this module I added a helper macro (so we can initialize the IPv4 address statically).

### CODE REVIEW

**Comment by kaspar030** [comment]
I was wondering when this discussion flames up. ;)

I strongly prefer using host byte order at the user API level. The overhead for changing the order as as soon as it is needed is negligible.

**Comment by miri64** [comment]
I find it weird for IPv4 addresses to be in host-byte order to be honest, but okay then I will change this macro, because I still think its useful for people who don't know about host- or network byteorder ;-)

**Comment by miri64** [comment]
> I find it weird for IPv4 addresses to be in host-byte order to be honest,

(especially since IPv6 addresses are in network byte order)

**Comment by miri64** [comment]
Mh come to think of it: excectly because of that, I would prefer to make the IPv4 address an array, too... What do you think?

**Comment by kaspar030** [comment]
> Mh come to think of it: excectly because of that, I would prefer to make the IPv4 address an array, too... What do you think?

Yes! I thought about that, too. Let's do that.

**Comment by miri64** [comment]
Done, but kept the helper macro, for people who like to work with `uint32_t` still ;-).

**Comment by miri64** [comment]
Arghs, also needs helper function for dynamic initialization to now...

**Comment by miri64** [comment]
Done

File: sys/include/net/sock.h [hunk 1]
```diff
@@ -172,7 +172,7 @@ typedef struct {
          */
         uint8_t ipv6[16];
 #endif
-        uint32_t ipv4;      /**< IPv4 address mode */
+        uint32_t ipv4;      /**< IPv4 address mode (in network byte order) */
```
Comment (kaspar030) [inline sys/include/net/sock.h]: Didn't we want to change this to be an array?

File: sys/include/net/sock.h [hunk 2]
```diff
@@ -222,9 +222,39 @@ struct _sock_tl_ep {
...
+
+/**
+ * @brief   Helper macro to initialize IPv4 address from integer dynamically.
+ *
+ * @param[in] ipv4  An IPv4 address array (needs to be of length 4 at minimum)
+ * @param[in] addr  The IPv4 address IPv4 address as 32-bit integer
+ */
+static inline void sock_init_ipv4(uint8_t *ipv4, uint32_t addr)
+{
+    const uint8_t a[] = SOCK_INIT_IPV4(addr);
```
Comment (kaspar030) [inline sys/include/net/sock.h]: This is not gonna work, the macro expects 4 arguments.
I actually think we can get rid of this function.

Users will usually not supply IPv4 addresses as "uint32".
They either get it from other functions (e.g., dns reply), or hardcode it.
In that case

```
ep = {.addr.ipv4= { a,b,c,d }}
```

would work as expected if the field is an uint8_t array, right?

**Comment by miri64** [comment]
Sorry there seems to be some commit fuckup...

File: sys/include/net/sock.h (hunk 2 above)
Comment (miri64) [inline sys/include/net/sock.h]: > Users will usually not supply IPv4 addresses as "uint32".
They either get it from other functions (e.g., dns reply), or hardcode it.
In that case

But a socket wrapper will ;-)

**Comment by miri64** [comment]
Now it's like I intended it to be

File: sys/include/net/sock.h (hunk 2 above)
Comment (kaspar030) [inline sys/include/net/sock.h]: example!

Comment (kaspar030) [inline sys/include/net/sock.h]: I mean, the output of ```inet_pton()``` or any ```struct in_addr``` can directly be copied into ```.addr.ipv4```, right?

Comment (miri64) [inline sys/include/net/sock.h]: True... I was more thinking in the direction of e.g.

```C
sock_init_ipv4(out->addr.ipv4, in_addr->sin_addr.s_addr);
```

But I guess you can do this with `memcpy()`, too.

**Comment by miri64** [comment]
@kaspar030 ping?

**Comment by kaspar030** [comment]
IMHO we don't need the helper macro and that static inline convenience function.

**Comment by miri64** [comment]
Done.

**Comment by kaspar030** [comment]
ACK, please squash!

I'll be on holidays for the next ten days, so unfortunately I have to reassign.

**Comment by miri64** [comment]
Squashed

**Comment by miri64** [comment]
Fixed a rebase error.

**Comment by kaspar030** [comment]
I'm just realizing that this complicates some things, e.g.:

```sock_udp_ep_t src = { .family = AF_INET, .addr.ipv4=INADDR_ANY};``` doesn't work anymore, also ```if  (dst->addr.ipv4 == 0xFFFFFFFF)``` ...

**Comment by miri64** [comment]
Well you wanted the macros removed. I was expecting something like that :-/.

**Comment by kaspar030** [comment]
How about we add a field ```uint32_t ipv4_u32``` to the union?

**Comment by miri64** [comment]
This leaves the byte-order unclear/inconsistent with the port.

**Comment by miri64** [comment]
To clarify:

* unclear: the type alone would suggest host byte-order, but the `.addr.ipv4` (and the way you plan to use it) sets it to network-byte order
* inconsistent: uint16_t for ports is in host byte-order, but uint32_t for address should be in network byte-order?

**Comment by kaspar030** [comment]
Hmm. I don't think that the byte order for an address matters that much, compared to a port number.

I find it absolutely more natural to write ```port = 25``` than ```ipv4 = 0xc0a80101```. The constants we're using (e.g., INADDR_ANY or INADDR_BROADCAST) would have to be (where applicable) in network byte order. And setting an address is probably done either be by string (```inet_pton(ep->addr.ipv4, "192.168.1.1", AF_INET)```), as result as one of our helper functions or just ```ep->addr.ipv4={192,168,1,1}```, which avoids the byte order matter.

What we gain by introducing an uint32_t version of the IPv4 address is the ability to do 32bit register compares due to 4b alignment, which, apart from being easier to code (and read), is probably way more efficient than having to do memcpys and memcmps.

**Pull request merged**